tests/integration/tls/helpers_manual_tls.py

```diff
@@ -0,0 +1,209 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import asyncio
+import base64
+import json
+import logging
+from collections import deque
+from typing import TYPE_CHECKING, NamedTuple
+
+from charms.tls_certificates_interface.v3.tls_certificates import (
+    generate_ca,
+    generate_certificate,
+    generate_private_key,
+)
+from juju.model import Model
+from juju.unit import Unit
+from tenacity import retry, stop_after_attempt, wait_exponential
+
+if TYPE_CHECKING:
+    from juju.action import Action
+
+logger = logging.getLogger(__name__)
+
+MANUAL_TLS_CERTIFICATES_APP_NAME = "manual-tls-certificates"
+
+
+class GettingOutstandingCertificateRequestsFailedError(Exception):
+    """Exception raised when getting outstanding certificate requests fails."""
+
+    def __init__(self, message: str) -> None:
+        """Initialise the exception."""
+        super().__init__(message)
+
+
+class CSRsMissingError(Exception):
+    """Exception raised when the number of CSRs in the queue is less than the expected number."""
+
+    def __init__(self, message: str) -> None:
+        """Initialise the exception."""
+        super().__init__(message)
+
+
+class ProvidingCertificateFailedError(Exception):
+    """Exception raised when providing a certificate fails."""
+
+    def __init__(self, message: str) -> None:
+        """Initialise the exception."""
+        super().__init__(message)
+
+
+class CSR(NamedTuple):
```

Comment:
Can you please explain? I used `NamedTuple` because I'm using CSRs as immutable. Based on my knowledge `NamedTuple` is more adequate and efficient than `Dataclass` for this use.